test/common/http/http2/codec_impl_test.cc

```diff
@@ -896,6 +896,64 @@ TEST_P(Http2CodecImplTest, EncodeMetadataWhileDispatchingTest) {
   EXPECT_CALL(response_decoder_, decodeMetadata_(_)).Times(size);
   request_encoder_->encodeHeaders(request_headers, true);
 }
+
+// Validate the keepalive PINGs are sent and received correctly.
```

Comment:
The comment is not correct, since the PING frame is not delivered to the peer in this test.
To make PING frames flow you will need to mock the `write` method on the `server_connection_`, like so:
```C++
  ON_CALL(server_connection_, write(_, _))
      .WillByDefault(
          Invoke([&](Buffer::Instance& data, bool) -> void { 
            // Peek into the data to check if this is indeed a PING frame
            client_wrapper_.buffer_.add(data); 
          }));
```

You can then dispatch the PING frame to the client codec like so `client_wrapper_.dispatch(Buffer::OwnedImpl(), *client_);`
Client codec will produce PING ACK in response and you can send it back to the server with the `client_->sendPendingFrames()` method and check that the ACK timeout timer was disabled.

Reply:
When I started writing the test, I expected the behavior you described. But it is definitely sending a PING and receiving an ACK. Validated in the debugger. Which is why the expectations below are met.

Reply:
Traced cause back to line 160 of this file:

```
  void setupDefaultConnectionMocks() {
    ON_CALL(client_connection_, write(_, _))
        .WillByDefault(Invoke([&](Buffer::Instance& data, bool) -> void {
          if (corrupt_metadata_frame_) {
            corruptMetadataFramePayload(data);
          }
          auto status = server_wrapper_.dispatch(data, *server_);
          if (!status.ok()) {
            throw ServerCodecError(std::move(status));
          }
        }));
    ON_CALL(server_connection_, write(_, _))
        .WillByDefault(Invoke([&](Buffer::Instance& data, bool) -> void {
          auto status = client_wrapper_.dispatch(data, *client_);
          if (!status.ok()) {
            throw ClientCodecError(std::move(status));
          }
        }));
  }
```